pkg/controller/provider/model/vsphere/tree.go

```diff
@@ -74,181 +66,67 @@ func (r *Tree) Build() (*TreeNode, error) {
-			}
-			if !r.Flatten {
-				node = child
-				defer func() {
-					node = node.Parent
-				}()
-			}
+			node.Children = append(node.Children, child)
+			node = child
+			defer func() {
+				node = node.Parent
+			}()
 		}
-		switch kind {
-		case FolderKind:
-			folder := model.(*Folder)
-		next:
-			for _, ref := range folder.Children {
-				switch r.Leaf {
-				case FolderKind:
-					if ref.Kind != r.Leaf {
-						continue next
-					}
-				}
-				m, err := r.getRef(ref)
-				if err != nil {
-					if errors.As(err, &InvalidRefError{}) {
-						continue
-					}
-					return liberr.Wrap(err)
-				}
-				err = walk(m, true)
-				if err != nil {
-					return liberr.Wrap(err)
-				}
-			}
-		case DatacenterKind:
-			var ref Ref
-			dc := model.(*Datacenter)
-			switch r.Leaf {
-			case ClusterKind, HostKind:
-				ref = dc.Clusters
-			case VmKind:
-				ref = dc.Vms
-			case NetKind:
-				ref = dc.Networks
-			case DsKind:
-				ref = dc.Datastores
-			case DatacenterKind:
-				// Leaf
-			default:
-				return InvalidKindError{kind}
-			}
-			m, err := r.getRef(ref)
+		for _, ref := range navigator(model) {
+			m, err := ref.Get(r.DB)
 			if err != nil {
-				if errors.As(err, &InvalidRefError{}) {
-					return nil
-				}
 				return liberr.Wrap(err)
 			}
 			err = walk(m, true)
 			if err != nil {
 				return liberr.Wrap(err)
 			}
-		case ClusterKind:
-			refList := []Ref{}
-			cluster := model.(*Cluster)
-			switch r.Leaf {
-			case HostKind, VmKind:
-				refList = cluster.Hosts
-			case NetKind:
-				refList = cluster.Networks
-			case DsKind:
-				refList = cluster.Datastores
-			case ClusterKind:
-				// Leaf
-			default:
-				return InvalidKindError{kind}
-			}
-			for _, ref := range refList {
-				m, err := r.getRef(ref)
-				if err != nil {
-					if errors.As(err, &InvalidRefError{}) {
-						return nil
-					}
-					return liberr.Wrap(err)
-				}
-				err = walk(m, true)
-				if err != nil {
-					return liberr.Wrap(err)
-				}
-			}
-		case HostKind:
-			refList := []Ref{}
-			host := model.(*Host)
-			switch r.Leaf {
-			case VmKind:
-				refList = host.Vms
-			case NetKind:
-				refList = host.Networks
-			case DsKind:
-				refList = host.Datastores
-			case HostKind:
-				// Leaf
-			default:
-				return InvalidKindError{kind}
-			}
-			for _, ref := range refList {
-				m, err := r.getRef(ref)
-				if err != nil {
-					if errors.As(err, &InvalidRefError{}) {
-						return nil
-					}
-					return liberr.Wrap(err)
-				}
-				err = walk(m, true)
-				if err != nil {
-					return liberr.Wrap(err)
-				}
-			}
-		case VmKind:
-			// Leaf
-		case NetKind:
-			// Leaf
-		case DsKind:
-			// Leaf
-		default:
-			return InvalidKindError{kind}
 		}
 
 		return nil
 	}
-	err := walk(r.Root, false)
+	err := walk(root, false)
 	if err != nil {
 		return nil, liberr.Wrap(err)
 	}
 
-	return root, nil
+	return treeRoot, nil
 }
 
 //
-// Get referenced model.
-func (r *Tree) getRef(ref Ref) (model Model, err error) {
```

Comment:
The `Get` functionality moved to the `Ref` object itself.  More generally useful.